Intro to Python - Lesson 1

mezarque opened the pull request:
This PR contains just the files needed for Intro to Python - Lesson 1. For a full overview of the lessons, see #48.

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
```diff
@@ -0,0 +1,548 @@
+
+Follow along in **Binder** by clicking the badge below:  
+
+[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/Arcadia-Science/arcadia-computational-training/das/pytutorial?labpath=docs%2Farcadia-users-group%2F20230228-intro-to-python-1%2Flesson.ipynb)
```
should the binder open to the launch screen so learners can open a blank notebook? I often find that coding along is really useful -- so instead of having everything pre-filled, each learner would type each cell as you type them

mezarque replied:
That make sense! I'll look into how to do that.

mezarque replied:
I added a `playground.ipynb` which links to the `lesson.ipynb` as the default notebook for Binder!  

I also updated the Binder link so it should point to where the `playground.ipynb` will live in `main` once this is merged. Hopefully it works but I can't test it quite yet!

taylorreiter replied:
ok we can check right after this is merged and patch as necessary

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
```diff
@@ -0,0 +1,548 @@
+---
+### 0.1 Jupyter Notebooks
+
+For the first workshop, we'll work in an interactive computing environment called a Jupyter notebook. To run a "cell" of code in Jupyter, you'll simply press <kbd>Shift</kbd>+<kbd>Return</kbd> from inside the cell. The result of the cell - the output - will display just below the cell.
```
```suggestion
For the first workshop, we'll work in an interactive computing environment called a Jupyter notebook. To run a "cell" of code in Jupyter, you'll press <kbd>Shift</kbd>+<kbd>Return</kbd> from inside the cell. The result of the cell - the output - will display just below the cell.
```

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
```diff
@@ -0,0 +1,548 @@
+
+> **Fun fact!**
+>
+> The Python programming language is not named for snakes, but instead for Monty Python, a signifier of the idea that the language should be [fun to use](https://en.wikipedia.org/wiki/Python_(programming_language)#Design_philosophy_and_features).  
```
love this!

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
```diff
@@ -0,0 +1,548 @@
+In general, you should expect to encounter *lots* of these errors, or "bugs", when you start getting into programming. **This is totally normal!**  
+In fact, probably 50% of programming is just squashing bugs.  
+
+If you're ever confused about the source of an error, websites such as [StackOverflow](https://stackoverflow.com/) are great places to look for answers.
```
```suggestion
If you're ever confused about the source of an error, websites such as [StackOverflow](https://stackoverflow.com/) are great places to look for answers.
One skill in learning to programming is learning what part of an error message to Google.
```

Or something like this -- indicating that a good step is googling the error message, and that it is a skill to figure out which part is the "right" part to google

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
```diff
@@ -0,0 +1,548 @@
+---
+## 3. Using variables
+
+You assign variables in Python using the `variable = value` syntax. You can perform operations on a mix of hard-coded values and variables interchangeably. This is very useful when you expect to use a variable in many different places in your script - Python keeps track of its value so you don't have to.
```
```suggestion
You assign variables in Python using the `variable = value` syntax. You can perform operations on a mix of hard-coded values and variables interchangeably. This is very useful when you expect to use a variable in many different places in your script - Python keeps track of its value so you don't have to. If you use variable names that correspond to the value the variable holds (think `wind_mph` for wind speed), this can also help make your code more readable for yourself and others.
```

mertcelebi commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.ipynb:
```diff
@@ -0,0 +1,884 @@
+{
```
<p>I'd probably add a note about what version of Python we're using.</p><p> </p><p>Also is it possible to have this lesson in the Python shell? And do you plan to eventually show people ways to call python files from the shell? I don't want people to assume you can only do Jupyter notebooks.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Arcadia-Science/arcadia-computational-training/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

mezarque replied:
<p>I thought it would be a little easier for people to get comfortable with the immediacy of interactive computing - I wasn't planning on covering python calling from the shell but it's definitely important to learn. Maybe I can pin it to the end of this lesson? The other two lessons are pretty jam-packed with information. I'll write something up and also try to thread it through the next few lessons.</p>

mezarque replied:
I added a section on running scripts from the command line, and changed the final problem in the problem set to have them write a script instead of a function!

mertcelebi commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.ipynb:
```diff
@@ -0,0 +1,884 @@
+{
```
<p>I'd probably add a note that double-quotes are preferable since that's what Black the linting tool recommends.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Arcadia-Science/arcadia-computational-training/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

mertcelebi replied:
<p>"Intuitive" is questionable but okay! :) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Arcadia-Science/arcadia-computational-training/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

mertcelebi replied:
<blockquote>In fact, probably 50% of programming is just squashing bugs.   </blockquote><p>I'd probably remove this. Doesn't add much and I don't want people to take it literally.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Arcadia-Science/arcadia-computational-training/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

mertcelebi replied:
<p>"exemplified"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Arcadia-Science/arcadia-computational-training/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

mertcelebi replied:
<p>do you mean variable <code>robot_name</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Arcadia-Science/arcadia-computational-training/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='43'/>

mertcelebi commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.ipynb:
```diff
@@ -0,0 +1,884 @@
+{
```
<p>What is a function?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Arcadia-Science/arcadia-computational-training/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

mezarque replied:
I added a definition of function to the new version.

mertcelebi commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.ipynb:
```diff
@@ -0,0 +1,884 @@
+{
```
<p>I think it's hard to show the value of functions here with just hello_world. So maybe in a later section, I'd do a more complicated function that allows you to abstract away some complexity. Maybe it could be a point a to point b distance calculation or something. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Arcadia-Science/arcadia-computational-training/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

mezarque replied:
I think some of that hopefully will come through in the problem set!

mertcelebi commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.ipynb:
```diff
@@ -0,0 +1,884 @@
+{
```
<p>I'd note that this is quite unusual and uncommon for languages</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Arcadia-Science/arcadia-computational-training/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

mezarque replied:
I made a note about this!

mertcelebi commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.ipynb:
```diff
@@ -0,0 +1,884 @@
+{
```
<p>What about block comments with """?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Arcadia-Science/arcadia-computational-training/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

mezarque replied:
I added a block comment example!

mertcelebi commented on docs/arcadia-users-group/20230228-intro-to-python-1/problem_set_1.ipynb:
```diff
@@ -0,0 +1,588 @@
+{
```
<p>What is the goal here? The ability for someone else to read a piece of code? Or do you actually care about the "efficiency" of it? Or do you think this code is not readable enough and making a couple changes would make it more readable?</p><p> </p><p>Depending on the goal, this may be not worthwhile for a beginner. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Arcadia-Science/arcadia-computational-training/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

mezarque replied:
I'm trying to accomplish a couple of things, perhaps not too effectively here: provide practice with reading and understanding code they didn't write, understanding enough to refactor it, and challenging them to write the code in both a more efficient and more readable way. I'll omit this problem for now, as perhaps those things are easier done with practice.

mertcelebi commented on docs/arcadia-users-group/20230228-intro-to-python-1/problem_set_1_answers.ipynb:
```diff
@@ -0,0 +1,607 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #5.    def factor(a, b, c):
</pre><p>Is this the solution? Doesn't seem different</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Arcadia-Science/arcadia-computational-training/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

mezarque replied:
Whoops! I'm going to remove this question, but I definitely forgot to save the solution.

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
````diff
@@ -0,0 +1,548 @@
+    10
+
+
+We can call a function from within another function call. For example, the `type()` function gives the type of a Python value.
+
+
+```python
+print(type(4))
+print(type('Hello, world!'))
+```
+
+    <class 'int'>
+    <class 'str'>
+
+
+---
````
do you use this much when you're programming? do you think it's best practice, or important for learner's mental model? if not it might be useful to take it out

mezarque replied:
I actually use this a lot when dealing with imported data in `pandas` - it helps manage NaNs and situations where columns have mixed data types!

mertcelebi commented on environment.yml:
```diff
@@ -5,3 +5,5 @@ channels:
 dependencies:
   - mkdocs=1.3.0
   - mkdocs-material=8.5.2
+  - matplotlib
```
It probably makes sense to have these be part of a separate env file.

mezarque replied:
What's the best way to do that and also keep the ability to create a Binder for this repo? I tried looking on the Binder reference but it isn't clear to me how/where I could specify a different `environment.yml` file... @taylorreiter would you happen to know how to do this?

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
```diff
@@ -0,0 +1,548 @@
+But there are plenty of cases where typecasting won't work, so be aware that this isn't a perfect solution.
+
+---
+## P1. Practice
```
what's your plan for doing this live? I sometimes:
1. make a powerpoint that has the questions on it so learners can see them
2. have the questions in a separate tab so I can copy and paste themonto my screen and into a learner chat (like our slack channel or zoom chat)

lmk if you want to brainstorm this or practice it, it can be hard managing this!

mezarque replied:
I can make a Google Slides document with the practice questions and examples!

mezarque replied:
I've made a Google Slides document [here](https://docs.google.com/presentation/d/13HoulLCuudnAgTJ03yMZUw_LWtB805SLF5o-tZUXtFo/edit?usp=sharing) that covers the lesson - I'm still working on the animations.

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
```diff
@@ -0,0 +1,548 @@
+
+<details>
+    <summary> Practice 2 Answer </summary>
+    This is a trick question! You might have expected the cell below to throw a <code>TypeError</code> because we're multiplying an <code>int</code> with a <code>str</code>. But <code>`4 * '15'`</code> is actually a special operation in Python; it's telling Python to repeat the string <code>'15'</code> four times. Therefore, the correct answer is <code>'15151515'</code>
```
love this example

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
```diff
@@ -0,0 +1,548 @@
+
+
+
+    'Hello Guido!'
```
lol

mertcelebi reviewed (changes requested):
Left a bunch of comments. Sadly I made those on the Jupyter notebook instead of the .md file..

One note, I'd add is:
* We should probably link to other accessible and good online resources. I like Replit's: https://replit.com/learn/100-days-of-python

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
```diff
@@ -0,0 +1,548 @@
+## 8. Comments
+
+A crucial part of writing good code is being able to understand what it does later.  
+For example, consider how mystifying it can be to try to read the code block below.
```
mystifying! great word!

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
```diff
@@ -0,0 +1,548 @@
+If you ended up writing the function above, it might be helpful to be able to know what the overall goal of the code is, and perhaps what each step is doing.
+
+This is where **comments** come to the rescue. Comments in Python begin on the left with a hash mark (`#`) and are ignored by the Python interpreter. You can even include a comment in the same line as functional code!  
+Consider the code below with comments - how does the code feel with comments and additional line breaks included?
```
this is such a nice demonstration

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
```diff
@@ -0,0 +1,548 @@
+---
+## 9. Problem Set
+
+To really get familiar with coding in Python (or any other language), the best way is simply to **write a lot of code**.  
```
```suggestion
To really get familiar with coding in Python (or any other language), the best way is to **write a lot of code**.  
```

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/lesson.md:
```diff
@@ -0,0 +1,548 @@
+CoLab works just like a Jupyter notebook, but it automatically saves your changes and keeps them as a cloud document.  
+**To get started on the problem set, make sure to first <mark>click "File" > "Save a copy in Drive"</mark> to make sure you don't lose your progress!**
+
+We'll have office hours throughout the week for you to stop by and work on your problem set with other Arcadians as well as to work through any bugs in your code.
```
```suggestion
We'll have AUG office hours next the week for you to stop by and work on your problem set with other Arcadians as well as to work through any bugs in your code.
Feel free to ping the #software-questions slack channel for anything that comes up in the meantime
```

is this ok?

mezarque replied:
Looks good!

taylorreiter commented on environment.yml:
```diff
@@ -5,3 +5,5 @@ channels:
   - mkdocs=1.3.0
   - mkdocs-material=8.5.2
+  - matplotlib
+  - pandas
```
nit: missing end of file new line

can you specify versions for  matplotlib and pandas as well to help future proof this?

mezarque replied:
I added versions and a newline!

taylorreiter reviewed (approved):
You really knocked this out of the park Dennis! this is really great. I left some comments/questions throughout, it would be great if you could address some of them (versions in yaml, removing "simply") and then feel free to reject the other suggestions. I made the suggestions in the lesson.md -- i hope it's not to annoying to backport to the ipynb...just realized while typing this that I should have done the whole review on NBreview...will do that moving forward

Other comments -- it might be good to say in the problem set 1 nb that there is an answer key and link to it. we'll also need to update the binder link to build from main, maybe we should do that before merge. 

If you wouldn't mind waiting for mert to look at this before merging, that would be good

mezarque commented:
Thanks for all your feedback! I've incorporated everything I could. I also made some lesson slides to use when people are working on the practice problems, linked again [here](https://docs.google.com/presentation/d/13HoulLCuudnAgTJ03yMZUw_LWtB805SLF5o-tZUXtFo/edit?usp=sharing).

I added a section about running Python scripts from the command line at the end - let me know if you think this makes enough sense.

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/hello_name.py:
```diff
@@ -0,0 +1,2 @@
+name = input("What's your name? ")
+print("Hello, " + name + "!")
```
missing EOF new line

taylorreiter commented on docs/arcadia-users-group/20230228-intro-to-python-1/hello_world.py:
```diff
@@ -0,0 +1 @@
+print("Hello world!")
```
missing EOF new line

taylorreiter reviewed (approved):
LGTM!
